File: Languages/python/algorithms-dsa/leetcode/trees/same_tree.py
"""
LeetCode #100 - Same Tree
Topic: Trees
Difficulty: Easy

Check if two binary trees are structurally identical using recursive comparison.

Time Complexity: O(n)
Space Complexity: O(h)
"""


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def is_same_tree(p: TreeNode | None, q: TreeNode | None) -> bool:
    if not p and not q:
        return True
    if not p or not q:
        return False
    return p.val == q.val and is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)


if __name__ == "__main__":
    t1 = TreeNode(1, TreeNode(2), TreeNode(3))
    t2 = TreeNode(1, TreeNode(2), TreeNode(3))
    assert is_same_tree(t1, t2) is True

    t3 = TreeNode(1, TreeNode(2))
    t4 = TreeNode(1, None, TreeNode(2))
    assert is_same_tree(t3, t4) is False

    assert is_same_tree(None, None) is True
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/trees/same_tree.cpp
/**
 * LeetCode 100: Same Tree
 * Topic: Trees
 * Difficulty: Easy
 *
 * Given the roots of two binary trees, check if they are the same.
 * Recursive comparison of structure and values.
 * Time: O(n), Space: O(h) where h = height
 */
#include <iostream>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool isSameTree(TreeNode* p, TreeNode* q) {
    if (!p && !q) return true;
    if (!p || !q) return false;
    return p->val == q->val &&
           isSameTree(p->left, q->left) &&
           isSameTree(p->right, q->right);
}

int main() {
    // Tree 1: [1, 2, 3]
    TreeNode* p1 = new TreeNode(1);
    p1->left = new TreeNode(2);
    p1->right = new TreeNode(3);

    TreeNode* q1 = new TreeNode(1);
    q1->left = new TreeNode(2);
    q1->right = new TreeNode(3);
    assert(isSameTree(p1, q1) == true);

    // Different structure
    TreeNode* p2 = new TreeNode(1);
    p2->left = new TreeNode(2);

    TreeNode* q2 = new TreeNode(1);
    q2->right = new TreeNode(2);
    assert(isSameTree(p2, q2) == false);

    // Both null
    assert(isSameTree(nullptr, nullptr) == true);

    cout << "All tests passed!" << endl;
    return 0;
}
